src/application.ts

```diff
@@ -44,14 +53,29 @@ export default class Application implements Owner {
     // Create ApplicationRegistry as a proxy to the underlying registry
     // that will only be available during `initialize`.
     let appRegistry = new ApplicationRegistry(this._registry, this.resolver);
-    this.initialize(appRegistry);
   }
 
-  initialize(registry: RegistryAccessor): void {
+  registerInstanceInitializer(initializer: InstanceInitializer) {
+    this._initializers.push(initializer);
+  }
+
+  initialize(): void {
+    let registry = this._registry;
+
     registry.register(`environment:/${this.rootName}/main/main`, Environment);
     registry.registerOption('template', 'instantiate', false);
+    registry.register(`document:/${this.rootName}/main/main`, document);
+    registry.registerOption('document', 'instantiate', false);
+    registry.registerInjection('environment', 'document', `document:/${this.rootName}/main/main`);
+    registry.registerInjection('component-manager', 'env', `environment:/${this.rootName}/main/main`);
+
+    let initializers = this._initializers;
+    for (let i = 0; i < initializers.length; i++) {
+      initializers[i].initialize(this);
```

Comment:
`initialize(registry)` is possible here, and removes the need for proxying to the registry from the `Application`